chromatic_fitting.py

```diff
@@ -78,127 +86,277 @@ def summarize_parameters(self):
+                    v.clear_prior()
+
     def initialize_empty_model(self):
         """
         Restart with an empty model.
         """
-        self.model = pm.Model()
+        self.pymc3_model = pm.Model()
 
     def attach_data(self, rainbow):
         """
         Connect a `chromatic` Rainbow dataset to this object.
         """
         self.data = rainbow._create_copy()
 
+
     def white_light_curve(self):
         """
         Generate inverse-variance weighted white light curve by binning Rainbow to one bin
         """
         self.white_light = self.data.bin(nwavelengths=self.data.nwave)
 
-    def choose_optimization_method(self,optimization_method='simultaneous'):
+    def choose_optimization_method(self, optimization_method='simultaneous'):
         """
         Choose the optimization method
+        [attach_data has to be run before!]
         """
         possible_optimization_methods = ["simultaneous", "white_light", "separate"]
         if optimization_method in possible_optimization_methods:
             self.optimization = optimization_method
+            if self.optimization == "separate":
+                try:
+                    self.create_multiple_models()
+                    self.change_all_priors_to_Wavelike()
+                except:
+                    pass
         else:
-            print("Unrecognised optimization method, please select one of: " + str(", ".join(possible_optimization_methods)))
+            print("Unrecognised optimization method, please select one of: " + str(
+                ", ".join(possible_optimization_methods)))
             self.optimization = "simultaneous"
 
-    def get_data(self):
+
+    def create_multiple_models(self):
+        """
+        Create a list of models to process wavelengths separately
+        """
+        self.pymc3_model = [pm.Model() for n in range(self.data.nwave)]
+        # if the LightCurve model is a CombinedModel then update the constituent models too
+        if isinstance(self, CombinedModel):
+            for mod in self.chromatic_models.values():
+                mod.pymc3_model = self.pymc3_model
+                mod.optimization = self.optimization
+
+
+    def change_all_priors_to_Wavelike(self):
+        for k, v in self.parameters.items():
+            if isinstance(v, Fitted) and not isinstance(v, WavelikeFitted):
+                self.parameters[k] = WavelikeFitted(v.distribution, **v.inputs)
+            if isinstance(v, Fixed) and not isinstance(v, WavelikeFixed):
+                self.parameters[k] = WavelikeFixed([v.value] * self.data.nwave)
+
+    def change_all_priors_to_notWavelike(self):
+        for k, v in self.parameters.items():
+            if isinstance(v, WavelikeFitted):
+                self.parameters[k] = Fitted(v.distribution, **v.inputs)
+            if isinstance(v, WavelikeFixed):
+                self.parameters[k] = Fixed(v.values[0])
+
+    def separate_wavelengths(self, i):
```

Comment:
Weirdly enough, I *think* you should be able to do exactly this with `self.data[i,:]`. I set up indexing Rainbows a while ago, and I think it basically does the same thing as this function? You might still want to keep this for flexibility, but it's just a heads up!

Reply:
self.data[i,:] doesn't seem to work for me, but maybe because I'm using an older version of chromatic?

Reply:
`b.data`
`Out[36]: <Simulated🌈(5w, 61t)>`
`b.data[0,:]`
`---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Input In [37], in <cell line: 1>()
----> 1 b.data[0,:]

File ~/Documents/Postdoc/CUBoulder/chromatic_fitting/chromatic/chromatic/rainbows/rainbow.py:805, in Rainbow.__getitem__(self, key)
    803         continue
    804     temporary = self.fluxlike[f][i_wavelength, :]
--> 805     new.fluxlike[f] = temporary[:, i_time]
    807 # finalize the new rainbow
    808 new._validate_core_dictionaries()

IndexError: too many indices for array: array is 1-dimensional, but 2 were indexed`

Reply:
Hmm, weird. Looks like this might be a bug that I didn't catch in how it handles single-wavelength Rainbows. I'll take a look in chromatic, but probably best to stick with your perfectly workable solution here!